### ISSUE TITLE
Spec compliance: clojure-lsp sends $/progress messages before initialize handshake completes

### ISSUE THREAD

**Issue by thomasjm** [clojure-lsp/clojure-lsp#1363]
**Describe the bug**

According to the LSP spec (v3.17):

> Until the server has responded to the initialize request with an [InitializeResult](https://microsoft.github.io//language-server-protocol/specifications/lsp/3.17/specification/#initializeResult), the client must not send any additional requests or notifications to the server. In addition the server is not allowed to send any requests or notifications to the client until it has responded with an [InitializeResult](https://microsoft.github.io//language-server-protocol/specifications/lsp/3.17/specification/#initializeResult), with the exception that during the initialize request the server is allowed to send the notifications window/showMessage, window/logMessage and [telemetry/event](https://microsoft.github.io//language-server-protocol/specifications/lsp/3.17/specification/#telemetry_event) as well as the window/showMessageRequest request to the client. In case the client sets up a progress token in the initialize params (e.g. property workDoneToken) the server is also allowed to use that token (and only that token) using the [$/progress](https://microsoft.github.io//language-server-protocol/specifications/lsp/3.17/specification/#workDoneProgressBegin) notification sent from the server to the client.

The rules are a little complicated, but I believe `clojure-lsp` is in violation because it sends `$/progress` notifications with a token of `clojure-lsp` before the initialization handshake has completed, even though the client hasn't set up such a progress token in the initialization params.

This surprises my LSP test harness (based on Haskell's [lsp-test](https://hackage.haskell.org/package/lsp-test)) and causes a crash.

**To Reproduce**

1. Obtain [Haskell Stack](https://docs.haskellstack.org/en/stable/) and also have Nix installed
2. `git clone --single-branch --branch clojure-lsp-crash git@github.com:codedownio/codedown-languages.git`
3. `cd codedown-languages/tests`
4. `stack run -- --print --clojure -f clojure-lsp`

**Expected behavior**

The below is from my test harness using Haskell's `lsp-test` package.

<details>
 <summary><b>clojure-lsp output</b></summary>
<pre>
--> {
    "id": 0,
    "jsonrpc": "2.0",
    "method": "initialize",
    "params": {
        "capabilities": {
            "general": {
                "markdown": {
                    "parser": ""
                },
                "regularExpressions": {
                    "engine": ""
                },
                "staleRequestSupport": {
                    "cancel": true,
                    "retryOnContentModified": []
                }
            },
            "textDocument": {
                "callHierarchy": {
                    "dynamicRegistration": true
                },
                "codeAction": {
                    "codeActionLiteralSupport": {
                        "codeActionKind": {
                            "valueSet": [
                                "quickfix",
                                "refactor",
                                "refactor.extract",
                                "refactor.inline",
                                "refactor.rewrite",
                                "source",
                                "source.organizeImports"
                            ]
                        }
                    },
                    "dataSupport": true,
                    "disabledSupport": true,
                    "dynamicRegistration": true,
                    "honorsChangeAnnotations": true,
                    "isPreferredSupport": true,
                    "resolveSupport": {
                        "properties": []
                    }
                },
                "codeLens": {
                    "dynamicRegistration": true
                },
                "colorProvider": {
                    "dynamicRegistration": true
                },
                "completion": {
                    "completionItem": {
                        "commitCharactersSupport": true,
                        "deprecatedSupport": true,
                        "documentationFormat": [
                            "plaintext",
                            "markdown"
                        ],
                        "insertReplaceSupport": true,
                        "insertTextModeSupport": {
                            "valueSet": []
                        },
                        "preselectSupport": true,
                        "resolveSupport": {
                            "properties": [
                                "documentation",
                                "details"
                            ]
                        },
                        "snippetSupport": true,
                        "tagSupport": {
                            "valueSet": [
                                1
                            ]
                        }
                    },
                    "completionItemKind": {
                        "valueSet": [
                            1,
                            2,
                            3,
                            4,
                            5,
                            6,
                            7,
                            8,
                            9,
                            10,
                            11,
                            12,
                            13,
                            14,
                            15,
                            16,
                            17,
                            18,
                            19,
                            20,
                            21,
                            22,
                            23,
                            24,
                            25
                        ]
                    },
                    "contextSupport": true,
                    "dynamicRegistration": true
                },
                "declaration": {
                    "dynamicRegistration": true,
                    "linkSupport": true
                },
                "definition": {
                    "dynamicRegistration": true,
                    "linkSupport": true
                },
                "documentHighlight": {
                    "dynamicRegistration": true
                },
                "documentLink": {
                    "dynamicRegistration": true,
                    "tooltipSupport": true
                },
                "documentSymbol": {
                    "dynamicRegistration": true,
                    "hierarchicalDocumentSymbolSupport": true,
                    "labelSupport": true,
                    "symbolKind": {
                        "valueSet": [
                            1,
                            2,
                            3,
                            4,
                            5,
                            6,
                            7,
                            8,
                            9,
                            10,
                            11,
                            12,
                            13,
                            14,
                            15,
                            16,
                            17,
                            18,
                            19,
                            20,
                            21,
                            22,
                            23,
                            24,
                            25,
                            26
                        ]
                    },
                    "tagSupport": {
                        "valueSet": [
                            1
                        ]
                    }
                },
                "foldingRange": {
                    "dynamicRegistration": true,
                    "lineFoldingOnly": false
                },
                "formatting": {
                    "dynamicRegistration": true
                },
                "hover": {
                    "contentFormat": [
                        "plaintext",
                        "markdown"
                    ],
                    "dynamicRegistration": true
                },
                "implementation": {
                    "dynamicRegistration": true,
                    "linkSupport": true
                },
                "onTypeFormatting": {
                    "dynamicRegistration": true
                },
                "publishDiagnostics": {
                    "relatedInformation": true,
                    "tagSupport": {
                        "valueSet": [
                            1,
                            2
                        ]
                    },
                    "versionSupport": true
                },
                "rangeFormatting": {
                    "dynamicRegistration": true
                },
                "references": {
                    "dynamicRegistration": true
                },
                "rename": {
                    "dynamicRegistration": true,
                    "onorsChangeAnnotations": true,
                    "prepareSupport": true,
                    "repareSupportDefaultBehavior": 1
                },
                "selectionRange": {
                    "dynamicRegistration": true
                },
                "semanticTokens": {
                    "dynamicRegistration": true,
                    "formats": [
                        "relative"
                    ],
                    "multilineTokenSupport": true,
                    "overlappingTokenSupport": true,
                    "requests": {
                        "full": {
                            "delta": true
                        },
                        "range": true
                    },
                    "tokenModifiers": [
                        "declaration",
                        "definition",
                        "readonly",
                        "static",
                        "deprecated",
                        "abstract",
                        "async",
                        "modification",
                        "documentation",
                        "defaultLibrary"
                    ],
                    "tokenTypes": [
                        "namespace",
                        "type",
                        "class",
                        "enum",
                        "interface",
                        "struct",
                        "typeParameter",
                        "parameter",
                        "variable",
                        "property",
                        "enumMember",
                        "event",
                        "function",
                        "method",
                        "macro",
                        "keyword",
                        "modifier",
                        "comment",
                        "string",
                        "number",
                        "regexp",
                        "operator"
                    ]
                },
                "signatureHelp": {
                    "dynamicRegistration": true,
                    "signatureInformation": {
                        "activeParameterSuport": true,
                        "documentationFormat": [
                            "plaintext",
                            "markdown"
                        ],
                        "parameterInformation": {
                            "labelOffsetSupport": true
                        }
                    }
                },
                "synchronization": {
                    "didSave": true,
                    "dynamicRegistration": true,
                    "willSave": true,
                    "willSaveWaitUntil": true
                },
                "typeDefinition": {
                    "dynamicRegistration": true,
                    "linkSupport": true
                }
            },
            "window": {
                "showDocument": {
                    "support": true
                },
                "showMessage": {},
                "workDoneProgress": true
            },
            "workspace": {
                "applyEdit": true,
                "configuration": true,
                "didChangeConfiguration": {
                    "dynamicRegistration": true
                },
                "didChangeWatchedFiles": {
                    "dynamicRegistration": true
                },
                "executeCommand": {
                    "dynamicRegistration": true
                },
                "semanticTokens": {
                    "refreshSupport": true
                },
                "symbol": {
                    "dynamicRegistration": true,
                    "symbolKind": {
                        "valueSet": [
                            1,
                            2,
                            3,
                            4,
                            5,
                            6,
                            7,
                            8,
                            9,
                            10,
                            11,
                            12,
                            13,
                            14,
                            15,
                            16,
                            17,
                            18,
                            19,
                            20,
                            21,
                            22,
                            23,
                            24,
                            25,
                            26
                        ]
                    },
                    "tagSupport": {
                        "valueSet": [
                            1
                        ]
                    }
                },
                "workspaceEdit": {
                    "changeAnnotationSupport": {
                        "roupsOnLabel": true
                    },
                    "documentChanges": true,
                    "normalizesLineEndings": true,
                    "resourceOperations": [
                        "create",
                        "delete",
                        "rename"
                    ]
                }
            }
        },
        "clientInfo": {
            "name": "lsp-test",
            "version": "0.14.1.0"
        },
        "processId": 3576890,
        "rootPath": "/home/tom/tools/codedown-languages/tests/test_runs/2022-11-03T02_12_55.522234408Z/results/0_command line options/Introduce parallel semaphore/Tests/Parallel/claim semaphore/Clojure/Clojure Nix environment/Jupyter runner/clojure-lsp: \"(foo 42)\"/clojure-lsp3576890",
        "rootUri": "file:///home/tom/tools/codedown-languages/tests/test_runs/2022-11-03T02_12_55.522234408Z/results/0_command%20line%20options/Introduce%20parallel%20semaphore/Tests/Parallel/claim%20semaphore/Clojure/Clojure%20Nix%20environment/Jupyter%20runner/clojure-lsp%3A%20%22%28foo%2042%29%22/clojure-lsp3576890",
        "trace": "off",
        "workspaceFolders": null
    }
}
<-- {
    "jsonrpc": "2.0",
    "method": "$/progress",
    "params": {
        "token": "clojure-lsp",
        "value": {
            "kind": "begin",
            "percentage": 0,
            "title": "clojure-lsp"
        }
    }
}
</pre>
</details>

**User details (please complete the following information):**
 - OS: Linux
 - Version: 
```
clojure-lsp 2022.10.05-16.39.51
clj-kondo 2022.10.05
```

**Comment by ericdallo** [comment]
Hey @thomasjm thanks for bringing this, I was kind of aware of that 😅 and not sure how we should improve that because it's a critical thing for clojure-lsp:
During `initialize`, clojure-lsp scan project and external jars which is the most expensive task of clojure-lsp, and this can take some seconds even minutes on huge projects, we have a fine tuned progress indication of that sent via `$/progress` and this ATM works pretty well with Emacs/vim but not with Vscode.
I'd love to improve this behavior supporting all editors and following the spec, but I don't want to lose current progress for users, WDYT?

c/c @mainej

**Comment by ericdallo** [comment]
I know `initialize` is a notification, but for clojure-lsp it doesn't make much sense to initialize the server without those analysis, that's why we hang that notification until everything is working properly

**Comment by thomasjm** [comment]
I think there's a simple way to be compliant: only enable the behavior if the initialize params contain a `workDoneToken` (and use that token value instead of `clojure-lsp`). Hopefully most editors have that capability already.

However, is it really necessary to do this expensive scan before returning the initialize response? Generally speaking the initialize response just tells the client about the capabilities of the server; hopefully it doesn't depend on the results of a scan. If you could return the initialize response immediately then it's fine to start sending progress messages after that.

**Comment by ericdallo** [comment]
> I think there's a simple way to be compliant: only enable the behavior if the initialize params contain a workDoneToken (and use that token value instead of "clojure-lsp"). Hopefully most editors have that capability already.

I think we already do that [here](https://github.com/clojure-lsp/clojure-lsp/blob/master/cli/src/clojure_lsp/server.clj#L499)

> However, is it really necessary to do this expensive scan before returning the initialize response? Generally speaking the initialize response just tells the client about the capabilities of the server; hopefully it doesn't depend on the results of a scan. If you could return the initialize response immediately then it's fine to start sending progress messages after that.

I thought about that, but I think we would fall in a state where client thinks it's ready to send requests like `definition`, `hover` etc, but server has no analysis about the code yet, making users think server is not working or something.

**Comment by thomasjm** [comment]
> I think we already do that [here](https://github.com/clojure-lsp/clojure-lsp/blob/master/cli/src/clojure_lsp/server.clj#L499)

Well if that were the case then there wouldn't be a problem :P. I think the problem is [here](https://github.com/clojure-lsp/clojure-lsp/blob/df7ff55d30c5dbeb0a3e28b7bcc21df6bf319a1e/cli/src/clojure_lsp/server.clj#L134), where the `publish-progress` method chooses a default value for `progress-token` when it is `nil`. Instead what it should do (at least in the initialization case) is not send the notification at all.

> I thought about that, but I think we would fall in a state where client thinks it's ready to send requests like definition, hover etc

That should be fine under the spec! The server is under no obligation to answer those requests until it's ready, and it can continue to provide startup feedback to the client using progress messages. The experience for the client should be the same, right? It should make no difference to me whether my `hover` request isn't showing up because it hasn't been sent yet or because it's blocked on the server.

**Comment by ericdallo** [comment]
> Well if that were the case then there wouldn't be a problem :P. I think the problem is [here](https://github.com/clojure-lsp/clojure-lsp/blob/df7ff55d30c5dbeb0a3e28b7bcc21df6bf319a1e/cli/src/clojure_lsp/server.clj#L134), where the publish-progress method chooses a default value for progress-token when it is nil. Instead what it should do (at least in the initialization case) is not send the notification at all.

I think must be a bug but we can change that as well.

> That should be fine under the spec! The server is under no obligation to answer those requests until it's ready, and it can continue to provide startup feedback to the client using progress messages. The experience for the client should be the same, right? It should make no difference to me whether my hover isn't showing up because it hasn't been sent yet or because it's blocked on the server.

Yes, in theory, but I don't think all editors will behave like that, we could do a POC in a branch and test that, it client UX is not affected I'm completely fine with that

**Comment by thomasjm** [comment]
:+1: I think the first change is the only one needed to be spec compliant. But the second one is needed to be "normal" :P

> Yes, in theory, but I don't think all editors will behave like that or server is completely ready for that

Not sure what you mean about the server being ready -- but I've been writing tests against lots of different LSP implementations and I promise that's how they all work, it should be fine for editors. For example, I use Haskell's LSP server all the time and I just confirmed this behavior--Haskell takes forever to index and process a large codebase, but it returns the initialize response immediately and then starts providing feedback via `$/progress` and `window/logMessage` as it works.

**Comment by ericdallo** [comment]
Good, I meant we have our own implementation of the LSP protocol on server side [lsp4clj](https://github.com/clojure-lsp/lsp4clj), so we might be no ready and need more changes, but sounds like the correct path.

I'll take a look later today to fix the token if there is a bug there

**Comment by ericdallo** [comment]
@thomasjm I made a commit adressing the publish of progress only if client provided the token, I'll keep this issue opened for the spec compliance

**Comment by thomasjm** [comment]
Thanks for the fast turnaround!

I usually build everything with Nix and I'm having a little trouble -- I noticed you're the maintainer on Nixpkgs so thought I'd ask here. It seems like the Nix derivation depends on downloading an external JAR from an existing GitHub release [here](https://github.com/NixOS/nixpkgs/blob/db3f2b35d30930b22730f7ead3eb1a1d7b9f4865/pkgs/development/tools/misc/clojure-lsp/default.nix#L14) ? I tried just replacing the `src` and deleting the `jar` attr, but got this error:

```
com.oracle.svm.driver.NativeImage$NativeImageError: /nix/store/i6rmgilcjh4sb6nvs631a5sq2hyw4g0f-source is a directory. (-jar requires a valid jarfile)
```

Any tips on how to build this? Thanks!

(BTW, the wiki link you sent has an outdated link to the old repo at `snoe/clojure-lsp`.)

**Comment by ericdallo** [comment]
@thomasjm, we recently added a flake at project root, check [here](https://clojure-lsp.io/installation/#nix), LMK if something is not working

**Comment by ericdallo** [comment]
Ops, we need to update the flake, we dropped `make` some weeks ago, we use `babashka` now via `bb cli-jar-for-native`

**Comment by thomasjm** [comment]
Hmm, I tried a little bit to put in the `bb` call but haven't gotten it to work. This is blocking my building it to test the progress messages fix, do you happen to have an ETA for the flake fix? Thanks!

**Comment by ericdallo** [comment]
Created a issue about the nix flake not working: https://github.com/clojure-lsp/clojure-lsp/issues/1373

**Comment by ericdallo** [comment]
Just a update on this, I quickly tried to start server immediately and then do the analysis and some things stopped working as I was guessing, so need to do a more careful change and understand that

**Comment by thomasjm** [comment]
Gotcha -- I'd be happy to test also, but it would be easier for me if I could do it using the flake. Any idea when that will be working?

**Comment by ericdallo** [comment]
The way we build the flake ATM uses https://github.com/jlesquembre/clj-nix and it doesn't support bb as a internal tool of the derivation, so we will need to wait for @jlesquembre help with that

**Comment by jlesquembre** [comment]
It's on my to-do list, but I didn't find time to work on it, sorry

**Comment by jlesquembre** [comment]
@thomasjm you can subscribe to https://github.com/jlesquembre/clj-nix/issues/37

**Comment by thomasjm** [comment]
@ericdallo I saw that the flake build has been fixed so I took a look at the latest version. It seems the extraneous `$/progress` messages have been fixed, presumably by 531e508f.

I think the server is LSP spec compliant now, so probably this issue can be closed? I know we were discussing another improvement involving a faster response of the `InitializeResult`. Not sure if any progress has been made on that, but I don't think it's essential. Maybe if anyone cares about it, it can be moved into its own issue.

I'll close this, feel free to reopen if I'm missing something.
